service-registry/src/main/java/io/servicecomb/serviceregistry/consumer/MicroserviceVersionRule.java

```diff
@@ -0,0 +1,145 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.servicecomb.serviceregistry.consumer;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.servicecomb.serviceregistry.api.registry.MicroserviceInstance;
+import io.servicecomb.serviceregistry.cache.InstanceCache;
+import io.servicecomb.serviceregistry.version.VersionRule;
+import io.servicecomb.serviceregistry.version.VersionRuleUtils;
+
+public class MicroserviceVersionRule {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MicroserviceVersionRule.class);
+
+  private final String appId;
+
+  private final String microserviceName;
+
+  private final VersionRule versionRule;
+
+  private MicroserviceVersion latestVersion;
+
+  // key is microserviceId
+  private Map<String, MicroserviceVersion> versions = new ConcurrentHashMap<>();
+
+  // key is instanceId
+  private Map<String, MicroserviceInstance> instances = new ConcurrentHashMap<>();
+
+  private InstanceCache instanceCache;
+
+  public MicroserviceVersionRule(String appId, String microserviceName, String strVersionRule) {
+    this.appId = appId;
+    this.microserviceName = microserviceName;
+    this.versionRule = VersionRuleUtils.getOrCreate(strVersionRule);
+
+    resetInstanceCache();
+  }
+
+  private void resetInstanceCache() {
+    instanceCache = new InstanceCache(appId, microserviceName, versionRule.getVersionRule(), instances);
+  }
+
+  public void addMicroserviceVersion(MicroserviceVersion microserviceVersion) {
+    if (!versionRule.isAccept(microserviceVersion.getVersion())) {
+      return;
+    }
+
+    versions.put(microserviceVersion.getMicroservice().getServiceId(), microserviceVersion);
+    resetLatest();
+
+    LOGGER.info("add microserviceVersion, appId={}, microserviceName={}, version={}, versionRule={}.",
+        microserviceVersion.getMicroservice().getAppId(),
+        microserviceVersion.getMicroservice().getServiceName(),
+        microserviceVersion.getVersion().getVersion(),
+        versionRule.getVersionRule());
+  }
+
+  public void deleteMicroserviceVersion(MicroserviceVersion microserviceVersion) {
+    if (!versionRule.isAccept(microserviceVersion.getVersion())) {
+      return;
+    }
+
+    if (versions.remove(microserviceVersion.getMicroservice().getServiceId()) == null) {
+      return;
+    }
+
+    resetLatest();
+    LOGGER.info("delete microserviceVersion, appId={}, microserviceName={}, version={}, versionRule={}.",
+        microserviceVersion.getMicroservice().getAppId(),
+        microserviceVersion.getMicroservice().getServiceName(),
+        microserviceVersion.getVersion().getVersion(),
+        versionRule.getVersionRule());
+  }
+
+  protected void resetLatest() {
```

Comment:
The name is better to be resetLatestVersion.

Reply:
done